File: templates/java/datastructures/SparseTable.java
package java.datastructures;

// This is the implementation of MIN Sparse Tree
// Try to use it in: 3599, 3605

// 1. Calculate log[i], where i: [1, n (arr.length)]
// a. log[1] = 0, log[i] = log[i / 2] + 1
// 2. Create Lookup Table [n][log[n] + 1]
// a. First row is equal to arr
// b. lookup[i][j] = f(lookup[i][j - 1], lookup[i + (1 << (j - 1))][j - 1]), where f is any idempotent function
// 3. Query
// a. Calculate intervalLen r - l + 1 and its log
// b. return f(lookup[l][j], lookup[r - (1 << j) + 1][j])

public class SparseTable {
    int n;
    int[] arr;
    int[] log;
    int[][] lookup;

    SparseTable(int[] arr) {
        this.n = arr.length;
        this.arr = arr;
        build();
    }

    void build() {
        this.log = new int[n + 1];
        log[1] = 0;
        for (int i = 2; i <= n; i++) {
            log[i] = log[i / 2] + 1;
        }

        this.lookup = new int[n][log[n] + 1];
        for (int i = 0; i < n; i++) {
            lookup[i][0] = arr[i];
        }
        for (int j = 1; j <= log[n]; j++) {
            for (int i = 0; i + (1 << j) - 1 < n; i++) {
                lookup[i][j] = Math.min(lookup[i][j - 1], lookup[i + (1 << (j - 1))][j - 1]);
            }
        }
    }

    int query(int l, int r) {
        int intervalLen = r - l + 1;
        int j = log[intervalLen];
        return Math.min(lookup[l][j], lookup[r - (1 << j) + 1][j]);
    }
}

File: templates/cpp/datastructures/SparseTable.cpp
#include <bits/stdc++.h>

using namespace std;

struct SparseTable {
    vector<vector<int>> lookup;
    vector<int> log;

    SparseTable(vector<int>& arr) {
        int n = arr.size();
        log.assign(n + 1, 0);
        for (int i = 2; i <= n; i++) log[i] = log[i / 2] + 1;
        int k = log[n - 1] + 1;

        lookup.assign(k, vector<int>(n, 0));
        copy(arr.begin(), arr.end(), lookup[0].begin());

        for (int i = 1; i < k; i++) {
            for (int j = 0; j + (1 << i) <= n; j++) {
                lookup[i][j] = min(lookup[i - 1][j], lookup[i - 1][j + (1 << (i - 1))]);
            } 
        }
    }

    int query(int l, int r) {
        int inv = log[r - l + 1];
        cout << inv << endl;
        return min(lookup[inv][l], lookup[inv][r - (1 << inv) + 1]); 
    }
};

int main() {
    // --- Test Case 1: Static RMQ ---
    vector<int> arr1 = {7, 2, 3, 0, 5, 10, 3, 12, 18};
    SparseTable st1(arr1);

    assert(st1.query(0, 8) == 0);
    assert(st1.query(0, 2) == 2);
    assert(st1.query(4, 7) == 3);
    assert(st1.query(5, 5) == 10);
    cout << "Test Case 1 Passed: Static RMQ verified." << endl;

    // --- Test Case 2: Overlapping Ranges ---
    vector<int> arr2 = {10, 8, 6, 4, 2, 1, 3, 5, 7, 9};
    SparseTable st2(arr2);

    assert(st2.query(0, 3) == 4); // [10, 8, 6, 4]
    assert(st2.query(2, 4) == 2); // [6, 4, 2]
    assert(st2.query(0, 9) == 1); // Full array
    cout << "Test Case 2 Passed: Overlapping range logic verified." << endl;

    cout << "\nAll Sparse Table tests passed!" << endl;
    return 0;
}